docs/source/async_api.rst

```diff
@@ -0,0 +1,497 @@
+
+.. note::
+
+    See https://neo4j.com/docs/operations-manual/current/configuration/ports/ for Neo4j ports.
+
+
+
+***********
+AsyncDriver
+***********
+
+Every Neo4j-backed application will require a :class:`neo4j.AsyncDriver` object.
+
+This object holds the details required to establish connections with a Neo4j database, including server URIs, credentials and other configuration.
+:class:`neo4j.AsyncDriver` objects hold a connection pool from which :class:`neo4j.AsyncSession` objects can borrow connections.
+Closing a driver will immediately shut down all connections in the pool.
+
+.. autoclass:: neo4j.AsyncDriver()
+   :members: session, close
+
+
+.. _async-driver-configuration-ref:
+
+Async Driver Configuration
+==========================
+
+:class:`neo4j.AsyncDriver` is configured exactly like :class:`neo4j.Driver`
+(see :ref:`driver-configuration-ref`). The only difference is that the async
+driver accepts an async custom resolver function:
+
+.. _async-resolver-ref:
+
+``resolver``
+------------
+A custom resolver function to resolve host and port values ahead of DNS resolution.
+This function is called with a 2-tuple of (host, port) and should return an iterable of 2-tuples (host, port).
+
+If no custom resolver function is supplied, the internal resolver moves straight to regular DNS resolution.
+
+The custom resolver function can but does not have to be a coroutine.
+
+For example:
+
+.. code-block:: python
+
+    from neo4j import AsyncGraphDatabase
+
+    async def custom_resolver(socket_address):
+        if socket_address == ("example.com", 9999):
+            yield "::1", 7687
+            yield "127.0.0.1", 7687
+        else:
+            from socket import gaierror
+            raise gaierror("Unexpected socket address %r" % socket_address)
+
+    # alternatively
+    def custom_resolver(socket_address):
+        ...
+
+    driver = AsyncGraphDatabase.driver("neo4j://example.com:9999",
+                                       auth=("neo4j", "password"),
+                                       resolver=custom_resolver)
+
+
+:Default: ``None``
+
+
+
+Driver Object Lifetime
+======================
+
+For general applications, it is recommended to create one top-level :class:`neo4j.AsyncDriver` object that lives for the lifetime of the application.
+
+For example:
+
+.. code-block:: python
+
+    from neo4j import AsyncGraphDatabase
+
+    class Application:
+
+        def __init__(self, uri, user, password)
+            self.driver = AsyncGraphDatabase.driver(uri, auth=(user, password))
+
+        async def close(self):
+            await self.driver.close()
+
+Connection details held by the :class:`neo4j.AsyncDriver` are immutable.
+Therefore if, for example, a password is changed, a replacement :class:`neo4j.AsyncDriver` object must be created.
+More than one :class:`.AsyncDriver` may be required if connections to multiple databases, or connections as multiple users, are required.
```

Comment:
maybe mention impersonation here, since multiple users can now be configured at the session level